aten/src/ATen/native/LinearCrossEntropy.cpp

```diff
@@ -0,0 +1,652 @@
+#define TORCH_ASSERT_ONLY_METHOD_OPERATORS
+#include <ATen/core/Tensor.h>
+#include <ATen/Dispatch.h>
+#include <ATen/native/Resize.h>
+#include <ATen/TensorIterator.h>
+#include <c10/util/irange.h>
+#include <limits>
+#include <tuple>
+#include <vector>
+#include <optional>
+
+#ifndef AT_PER_OPERATOR_HEADERS
+#include <ATen/Functions.h>
+#include <ATen/NativeFunctions.h>
+#else
+#include <ATen/ops/empty.h>
+#include <ATen/ops/linear.h>
+#include <ATen/ops/cross_entropy_loss.h>
+#include <ATen/ops/zeros.h>
+#include <ATen/ops/zeros_like.h>
+#include <ATen/ops/full.h>
+#include <ATen/ops/ones.h>
+#include <ATen/ops/max.h>
+#include <ATen/ops/exp.h>
+#include <ATen/ops/log.h>
+#include <ATen/ops/logsumexp.h>
+#include <ATen/ops/where.h>
+#include <ATen/ops/ge.h>
+#include <ATen/ops/lt.h>
+#include <ATen/ops/logical_and.h>
+#include <ATen/ops/logical_or.h>
+#include <ATen/ops/logical_not.h>
+#include <ATen/ops/masked_fill.h>
+#include <ATen/ops/sub.h>
+#include <ATen/ops/add.h>
+#include <ATen/ops/mul.h>
+#include <ATen/ops/index_select.h>
+#include <ATen/ops/gather.h>
+#include <ATen/ops/nonzero.h>
+#include <ATen/ops/maximum.h>
+#include <ATen/ops/gt.h>
+#include <ATen/ops/div.h>
+#include <ATen/ops/ne.h>
+#include <ATen/ops/sum.h>
+#include <ATen/ops/linear_cross_entropy_backward_native.h>
+#include <ATen/ops/linear_cross_entropy_native.h>
+#endif
+
+namespace at::native {
+
+// Strategy selection for optimal chunking approach
+enum class ChunkingStrategy {
+    NAIVE,           // No chunking - standard approach
+    VOCAB_CHUNKING,  // Chunk vocabulary dimension (existing)
+    BATCH_CHUNKING   // Chunk batch dimension (new)
+};
+
+Tensor batch_chunking_cpu(
+    const Tensor& input,
+    const Tensor& weight,
+    const Tensor& target,
+    const std::optional<Tensor>& bias_opt,
+    int64_t reduction,
+    int64_t ignore_index,
+    double label_smoothing);
+
+// Determine optimal chunking strategy based on input dimensions and user preference
+// Based on memory reduction analysis and empirical validation
+inline ChunkingStrategy select_chunking_strategy(
+    int64_t vocab_size,
+    int64_t flattened_batch_size,
+    c10::string_view strategy) {
+
+    if (strategy == "none") {
+        return ChunkingStrategy::NAIVE;
+    } else if (strategy == "vocab") {
+        return ChunkingStrategy::VOCAB_CHUNKING;
+    } else if (strategy == "batch") {
+        return ChunkingStrategy::BATCH_CHUNKING;
+    } else if (strategy == "auto") {
+        // Empirically validated chunk sizes for optimal memory/compute balance
```

Comment:
Where are these numbers from?

I'm not sure there is precedence for providing a heuristic as the default, would it make sense to make the user explicitly choose the strategy they want to use instead?